### PR TITLE
[query] Remove TableGroupWithinPartitions IR, Attempt 2

### PR DESCRIPTION
**Pull request by johnc1231**
There was a PruneDeadFields bug the first time I tried this, and I only just now went and fixed it. I can't reopen #9432 though because the branch was force pushed since I closed it. 

Anyway, like it says in #9432, we no longer need `TableGroupWithinPartitions`, as we have the ability to write it explicitly in Python now using a mix of `TableMapPartitions` and `StreamGrouped`.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 1]
```diff
@@ -415,7 +415,7 @@ object PruneDeadFields {
         val reqRowsType = TStream(requestedType.rowType)
         val bodyDep = memoizeValueIR(body, reqRowsType, memo)
         val depGlobalType = unifySeq(child.typ.globalType,
-          bodyDep.eval.lookupOption(gName).map(_.result()).getOrElse(Array()))
+          bodyDep.eval.lookupOption(gName).map(_.result()).getOrElse(Array()) :+ requestedType.globalType)
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala:418]: Also worth looking here. This was a bug I uncovered while doing this. We need to consider the requested type when doing the `PruneDeadFields` pass for `TableMapPartitions`.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala:418]: ahh, good

File: hail/python/hail/table.py [hunk 2]
```diff
@@ -3493,7 +3493,13 @@ def multi_way_zip_join(tables, data_field_name, global_field_name) -> 'Table':
             [t._tir for t in tables], data_field_name, global_field_name))
 
     def _group_within_partitions(self, name, n):
-        return Table(ir.TableGroupWithinPartitions(self._tir, name, n))
+        def grouping_func(part):
+            groups = part.grouped(n)
+            key_names = list(self.key)
+            return groups.map(lambda group:
+                              hl.struct(**{key_name: group[0][key_name] for key_name in key_names}).annotate(**{name: group}))
```
Comment (patrick-schultz) [inline hail/python/hail/table.py:3500]: This is probably better as `group[0].select(*key_names, **{name: group})` (I think that works)

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

File: hail/python/hail/table.py (hunk 2 above)
Comment (johnc1231) [reply hail/python/hail/table.py]: Nice, that is better.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**